Feat/object storage

Geun-Oh opened the pull request:
## Summary

Add Object Storage service (including bucket, bucket ACL, object and object ACL)

## File added

objectstorage_bucket_acl_test.go
objectstorage_bucket_acl.go
objectstorage_bucket_data_source_test.go
objectstorage_bucket_data_source.go
objectstorage_bucket_test.go
objectstorage_bucket.go
objectstorage_object_acl_test.go
objectstorage_object_acl.go
objectstorage_object_copy_test.go
objectstorage_object_copy.go
objectstorage_object_data_source_test.go
objectstorage_object_data_source.go
objectstorage_object_test.go
objectstorage_object.go

## Related Issue

TBD

## Error & Additional

- [x] 컨벤션 관련 사항 및 누락된 코드 확인 부탁드립니다. 수정완료 되면 data_source 코드까지 추가하여 본 pr 올려두겠습니다.
- [x] 현재 object의 경우 테스트 코드에서 랜덤한 버킷이 아닌 제 지정된 버킷에서 돌아가도록 되어 있습니다. 독립적이지 못한 부분이어서 수정이 필요한데, 해당 부분 적절한 방법 알려주시면 해당 방법대로 구현해두겠습니다. 현재는 테스팅하면서 랜덤한 버킷을 생성하여 객체를 업로드하는 방향을 생각 중에 있습니다.
- 스키마 관련해서 추가가 필요한 사항 말씀주시면 추가해두고, 이후 본 PR올리기 이전에 docs => Argument Reference 에 내용 채워두고자 
합니다.

Geun-Oh committed:
chore: Install AWS SDK for Object Storage API

Geun-Oh committed:
feat: Add S3 Client configuration

Geun-Oh committed:
feat: Add Object Storage Resources

Geun-Oh committed:
feat: Add objectstorage_bucket resource

Geun-Oh committed:
feat: Add objectstorage_bucket_acl resource

Geun-Oh committed:
feat: Add objectstorage_object resource

Geun-Oh committed:
feat: Add objectstorage_object_acl resource

Geun-Oh committed:
fix: add "_bucket" suffix to bucket resource

Geun-Oh committed:
feat: Add examples of object storage

Geun-Oh committed:
docs: Add initial documentation of object storage

Geun-Oh commented:
go test -v github.com/terraform-providers/terraform-provider-ncloud/internal/service/objectstorage
=== RUN   TestAccResourceNcloudObjectStorage_bucket_acl_basic
--- PASS: TestAccResourceNcloudObjectStorage_bucket_acl_basic (16.69s)
=== RUN   TestAccDataSourceNcloudObjectStorage_bucket_basic
--- PASS: TestAccDataSourceNcloudObjectStorage_bucket_basic (11.78s)
=== RUN   TestAccResourceNcloudObjectStorage_bucket_basic
--- PASS: TestAccResourceNcloudObjectStorage_bucket_basic (11.40s)
=== RUN   TestAccResourceNcloudObjectStorage_object_acl_basic
--- PASS: TestAccResourceNcloudObjectStorage_object_acl_basic (26.52s)
=== RUN   TestAccResourceNcloudObjectStorage_object_copy_basic
--- PASS: TestAccResourceNcloudObjectStorage_object_copy_basic (31.74s)
=== RUN   TestAccDataSourceNcloudObjectStorage_object_basic
--- PASS: TestAccDataSourceNcloudObjectStorage_object_basic (21.52s)
=== RUN   TestAccResourceNcloudObjectStorage_object_basic
--- PASS: TestAccResourceNcloudObjectStorage_object_basic (21.92s)

Geun-Oh commented:
@dusdjhyeon 빠진거없나 봐주삼

Geun-Oh committed:
feat: add type casting function for nested listValue

Geun-Oh committed:
feat: Add BucketRegion to bucket chema

Geun-Oh committed:
fix: Add detailed schema of grants & type casting method

Geun-Oh committed:
chore: Add error exception

Geun-Oh committed:
feat: Add regex validator for resource id

Geun-Oh committed:
chore: remove redundant EOL

Geun-Oh committed:
fix: Remove BucketName, BucketRegion from schema(doesn't provided)

Geun-Oh committed:
refactor: Add TrimForIDParsing util

Geun-Oh committed:
feat: Add temporary file creation logic

Geun-Oh committed:
feat: Add aclOptions for testing

Geun-Oh commented:
@wonchulee @youngmn Thanks for the review :)

wonchulee commented on internal/conn/api_client.go:
```diff
@@ -0,0 +1,34 @@
+
+	cfg, err := config.LoadDefaultConfig(context.TODO(),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(api.AccessKey, api.SecretKey, "")),
+		config.WithRegion("KR"),
```
우선 KR 리전을 대상으로 작업해주신 것으로 이해했습니다. 개발하시면서 configuration에 입력된 Region 정보를 바탕으로 region 및 region에 해당하는 endpoint를 설정하도록 추가해주면 좋을 것 같습니다!

Geun-Oh replied:
넵! 리전에 종속되지 않도록 구성해보겠습니다.

wonchulee commented on internal/conn/api_client.go:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	newClient := s3.NewFromConfig(cfg, func(o *s3.Options) {
+		o.BaseEndpoint = aws.String(endpoint)
```
(사소한 부분입니다) aws.String()과 동일한 ncloud.String()과 같은 helper 함수들이 있으니 해당 함수로 대체해주셔도 좋겠습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,264 @@
+		resp.Diagnostics.AddError("CREATING ERROR", "HeadBucket call failed")
+	}
+
+	err = waitBucketeCreated(ctx, o.config, plan.BucketName.String())
```
(사소한 부분) waitBucket`e`Created 중간에 오타가 섞인 것 같습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,264 @@
+				Validators: []validator.String{
+					stringvalidator.All(
+						stringvalidator.LengthBetween(3, 15),
+						stringvalidator.RegexMatches(regexp.MustCompile(`^[가-힣A-Za-z0-9-]+$`), "Allows only hangeuls, alphabets, numbers, hyphen (-)."),
```
KR 외의 지역에서 값을 입력할 때 validation 기준이 달라질 수 있을까요? (ex: 일본어)

Geun-Oh replied:
콘솔에서 확인되는 bucket name 기준에 맞게 수정하였습니다.

Geun-Oh replied:
해당 부분 콘솔을 통해 네이밍 규칙 확인하여 변경해두었습니다 :)

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,264 @@
+			for _, bucket := range output.Buckets {
+				if "\""+*bucket.Name+"\"" == bucketName {
+					return bucket, "created", nil
+				}
```
(사소한 부분) bucketName 비교시 "\""를 제외하고, 사전에 bucketName만 가져온 값으로 비교하면 좋을 것 같습니다. (이전 단계에서 Model 에 해당 값을 담을 때 제거한다거나..)

Geun-Oh replied:
해당 부분 추후에 문자열 파싱용으로 만든 `TrimForParsing` 유틸 활용해서 수정해두었습니다. 감사합니다 :)

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,264 @@
+func waitBucketeCreated(ctx context.Context, config *conn.ProviderConfig, bucketName string) error {
+	stateConf := &retry.StateChangeConf{
+		Pending: []string{"creating"},
+		Target:  []string{"created"},
```
(사소한 부분) Pending, Target에 사용되는 string literal 은 실수를 방지하기 위해 const 변수로 선언하여 사용하면 더 좋을 것 같습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,264 @@
+		return
+	}
+
+	o.ID = types.StringValue(fmt.Sprintf("https://%s.object.ncloudstorage.com/%s", strings.ToLower(config.RegionCode), bucketName))
```
id에 endpoint 주소를 함께 포함할 필요는 없을 것 같아요, 위에서 언급한 것처럼 사용자가 입력한 configuration 에 따라서 Region 정보를 받아오도록 변경한다면, id 값에 region 값도 함께 전달하지 않아도 region 별 스토리지 구분이 가능할 것 같습니다. 그럼 bucket을 식별할 수 있는 이름 정도로 id를 설정하면 어떨까요?

Geun-Oh replied:
아하 확인했습니다!

Geun-Oh replied:
넵!! bucket 이름으로 id 구성하도록 하여 초기 리전 설정 시 이름만으로 유일성 보장되도록 설정해두었습니다.

Geun-Oh committed:
fix: Removed Unavailable object acl

Geun-Oh committed:
feat: Add Bucket data source

Geun-Oh committed:
fix: Add new bucket creation for independant testing

Geun-Oh committed:
fix: aws.String => ncloud.String

Geun-Oh committed:
chore: fix typo

Geun-Oh committed:
chore: change string value with constants

Geun-Oh committed:
refactor: expand TrimForParsing to parse vairous strings (not only ID)

Geun-Oh committed:
feat: use region info from initial configuration

Geun-Oh committed:
refactor: Set bucket id same as bucket name

Geun-Oh commented:
- [x] TODO : 추후에 object_acl 의 경우도 특정 엔드포인트에 특정되지 않도록 테스트 구성하기

Geun-Oh committed:
fix: Add detailed string validator with regex for bucket_name

Geun-Oh committed:
feat: Add NewObjectDataSource data source

Geun-Oh committed:
refactor: Move ObjectIDParser to object.go

Geun-Oh committed:
refactor: Set Object Resource ID with new format

Geun-Oh commented:
Required Attribute는 정의되었으니, 병합한다면 이 정도에서 병합한 뒤 추후 필요한 속성이나 data source 들을 추가하는 과정을 거쳐도 괜찮을 것 같습니다.

Geun-Oh committed:
feat: set independent object testing

Geun-Oh committed:
feat: Add Owner info & Creation Date to bucket data source

Geun-Oh committed:
feat: Add OwnerID & OwnerDisplayname at object_acl attribute

Geun-Oh committed:
fix: set bucket_id & object_id w/o NRN

Geun-Oh committed:
docs: Add documentation of object storage resources

Geun-Oh committed:
feat: Change format - make resource id with '/' in object storage - object

Geun-Oh committed:
feat: Add objectstorage_object_copy

Geun-Oh committed:
fix: Change to do not use pointer method

Geun-Oh committed:
docs: Add data source documentation of bucket

Geun-Oh committed:
docs: Add object_storage_object.md in data-source

Geun-Oh committed:
docs: Add object_storage_object_copy.md

Geun-Oh committed:
feat: Add CreationDate attribute to bucket

Geun-Oh committed:
fix: fix logic to check whether output is nil

Geun-Oh committed:
fix: Add format of creationDate

wonchulee commented on internal/conn/config.go:
```diff
@@ -73,9 +77,12 @@ type NcloudAPIClient struct {
 }
 
 func (c *Config) Client() (*NcloudAPIClient, error) {
+	s3Endpoint := fmt.Sprintf("https://%s.object.ncloudstorage.com", strings.ToLower(c.Region))
```
민간 뿐만 아니라 gov, fin 환경의 endpoint 설정도 함께 고려해주시면 좋겠습니다.
그리고 내부적으로 베타 환경에서 테스트하는 경우 endpoint를 주입할 수 있는 방법이 제공되면 좋을 것 같은데요, 예를 들면 환경변수로 endpoint를 오버라이딩해서 사용할 수 있도록 제공하면 어떨까요?

Geun-Oh replied:
해당 부분 NCLOUD_ENDPOINT 라는 환경변수가 존재하는 경우 해당 변수를 최우선 endpoint로 설정하도록 작업해두었습니다 :)

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,290 @@
+		Pending: []string{CREATING},
+		Target:  []string{CREATED},
+		Refresh: func() (interface{}, string, error) {
+			output, err := config.Client.ObjectStorage.ListBuckets(ctx, &s3.ListBucketsInput{})
```
아마도 사전에 다른 방법을 확인해보고 사용하신 것일거라 생각됩니다만,
생성/삭제 등 작업 수행후 버킷의 상태를 확인하기 위해서 ListBuckets()으로 전체 버킷을 조회하는 대신 HeadBucket() 등의 다른 API를 사용하여 단일 버킷의 상태를 조회하도록 개선 가능할까요? (다른 리소스에서도 마찬가지)

Geun-Oh replied:
해당 부분의 경우 HeadBucket으로 wait을 처리하게 되면 bucket 정보를 불러오지 못하는 이슈가 있었습니다. 반면에 ListBuckets로 가져오는 경우 해당 버킷의 존재 여부를 확인할 수 있었기에 위와 같이 처리해두었습니다(HeadBucket이 캐싱된 내용을 참조하는 것으로 판단됨).

말씀주신 것처럼 ListBuckets를 수행하게되면 offset 도 주어지지 않는 API이기에 많은 리스트를 여러 번 호출해야한다는 문제가 있으나, 현재 가용한 수단을 적용한 점 있습니다.

Geun-Oh replied:
영민님께도 여쭤본 적이 있었는데, 우선은 ListBucket으로 가져오도록 처리를 진행한 뒤 추후 방법을 찾아보는 것을 말씀해주셔서 우선 진행했습니다.

이외에 HeadBucket을 사용 가능한 부분은 추가로 수정해두겠습니다!

Geun-Oh replied:
```
https://github.com/Geun-Oh/terraform-provider-ncloud/pull/1#issuecomment-2328138634

해당 사항으로 정리
```

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,290 @@
+			},
+			"creation_date": schema.StringAttribute{
+				Computed: true,
+				Optional: true,
```
`creation_date` attribute는 API로 반환된 결과를 사용하지만 사용자가 직접 업데이트하지 않는 케이스로 보여요, 그럼 optional 속성은 제거하면 좋겠습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,290 @@
+func (o *bucketResource) Schema(_ context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": framework.IDAttribute(),
```
구 SDK는 ID attribute가 자동으로 생성되어, 모든 리소스에 필수적으로 사용되었습니다. framework는 선택적으로 추가할 수 있습니다. id가 name과 동일한 값으로 사용하는 대신 attribute에서 삭제하고 bucket_name을 ImportState에서 활용하도록 변경하면 좋겠네요! (다른 리소스에서도 마찬가지)

Geun-Oh replied:
bucket의 경우 해당 사항을 기반으로 수정해두었습니다.
object의 경우 타 attr를 직접 가져다 사용할 수 없어 그대로 진행하고자 합니다.

Geun-Oh replied:
id관련 Attr를 삭제하는 경우 아래 에러를 지속적으로 마주하게 되는 것 같습니다..아마도 테스트 시 이전 버전 로직을 함께 돌기 때문이라고 생각이 되는데, 다시 확인해보고 변경 가능하면 변경해두겠습니다!

=== RUN   TestAccResourceNcloudObjectStorage_bucket_acl_basic
    testing_new_config.go:82: no "id" found in attributes
    testing_new.go:63: no "id" found in attributes
--- FAIL: TestAccResourceNcloudObjectStorage_bucket_acl_basic (12.55s)
=== RUN   TestAccDataSourceNcloudObjectStorage_bucket_basic
^Csignal: interrupt
FAIL	github.com/terraform-providers/terraform-provider-ncloud/internal/service/objectstorage	15.190s

Geun-Oh replied:
```
해당 부분 bucket의 경우 id가 불필요함을 확인함.

다만 아직 framework 자체 테스팅에서 이전 로직을 활용하고 있어 id를 포함해야 함을 확인함.

따라서 우선 이대로 진행하고, 추후 업데이트를 확인하여 수정하는 것으로 정리.
```

youngmn commented on internal/conn/config.go:
```diff
@@ -47,6 +50,7 @@ type Config struct {
 	AccessKey string
 	SecretKey string
 	Region    string
+	Endpoint  string
```
사용 되는 곳이 있을까요?

Geun-Oh replied:
해당 부분 초기 테스트 시 사용되었던 것이어서, 삭제하겠습니다!

youngmn commented on internal/conn/config.go:
```diff
@@ -73,9 +77,12 @@ type NcloudAPIClient struct {
 }
 
 func (c *Config) Client() (*NcloudAPIClient, error) {
+	s3Endpoint := fmt.Sprintf("https://%s.object.ncloudstorage.com", strings.ToLower(c.Region))
```
c.Region 에 입력되는 값은
민간 : KR, HK, SGN, JPN, USWN, DEN 입니다. 값에 따라 주소 변환하는 로직이 필요해 보이네요

공공은 https://kr.object.gov-ncloudstorage.com
https://krs.object.gov-ncloudstorage.com

금융은 https://kr.object.fin-ncloudstorage.com

Geun-Oh replied:
넵 확인했습니다!

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_acl.go:
```diff
@@ -0,0 +1,354 @@
+	}
+}
+
+func (b *bucketACLResource) Update(context.Context, resource.UpdateRequest, *resource.UpdateResponse) {
```
PutBucketAcl() 을 사용하여 acl 업데이트가 가능하다면 함께 추가해주시면 좋겠습니다

Geun-Oh replied:
해당 부분 기존 Create과 동일한 동작을 가져가므로 Terraform의 업데이트 로직에 맡기고(삭제 후 생성), 추가 구현은 하지 않았습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_acl.go:
```diff
@@ -0,0 +1,354 @@
+
+	tflog.Info(ctx, "PutObjectACL response="+common.MarshalUncheckedString(response))
+
+	if err := waitBucketACLApplied(ctx, b.config, bucketName); err != nil {
```
일부 리소스들은 적용에 시간이 소요되지 않고 즉시 확인 가능한 것들도 있어보이는데요, ACL도 PutBucketAcl() 후 즉시 반영되어 확인 가능(wait 불필요)한 리소스인지 더블 체크 해주시면 좋겠습니다 (bucket acl 외 다른 리소스 마찬가지)

Geun-Oh replied:
해당 부분은 적용 이후에 Get으로 가져와서 필요한 필드들을 채워주는(rule, grants 등의 정보) 것을 목적으로 추가적으로 Get을 수행하는 바 있습니다!

wonchulee commented on internal/service/objectstorage/objectstorage_object.go:
```diff
@@ -0,0 +1,322 @@
+		return
+	}
+
+	plan.refreshFromOutput(ctx, o.config)
```
(사소한부분) getobject을 refreshFromOutput() 내에서 사용하고 있는데, 다른 곳에서 사용하는 것과 다른 형태여서 getobject 후 refreshFromOutput을 수행해도 좋겠습니다만, 사소한 부분이라 반영하지 않아도 괜찮습니다. (누락된 부분인 줄 알았음)

Geun-Oh replied:
```
해당 부분 refreshFromOutput에서 Get operation을 수행하는 형태로 변경.
이후 타 리소스들 또한 해당 포맷을 유지하도록 변경.
```

wonchulee commented on internal/service/objectstorage/objectstorage_object_acl.go:
aws에서는 object의 attribute 중 하나로 acl을 제공하여 설정하고 있는 것으로 보입니다. 별도 리소스로 분리하는 것이 낫다고 판단하신 것 같습니다. 관련하여 의견 부탁드릴 수 있을까요?

Geun-Oh replied:
해당 부분은 기본적인 Terraform의 철학에 따라 각 리소스를 독립적으로 정의되고 관리되어야한다고 생각하여 진행했습니다.
또한, 두 리소스의 ACL 관리 방식이 통일되는 것이 사용자 입장에서 이해하고 사용하기 좋을 것이라고 판단하였습니다.

다만 AWS에서 왜 Bucket은 별도의 ACL리소스를 두고, Object에는 왜 하나의 리소스 내에서 ACL을 관리하도록 하였는지 좀 더 명확한 이유를 찾아보고자 합니다.

Geun-Oh replied:
```
해당 부분 AWS에서의 추가적인 이유 파악 어려움.
추후 N:N 등의 추가 관계 설정 등의 feature 추가 가능성을 고려하여 독립적인 리소스로 관리.

타당한 이유를 찾으면 해당 이유 기반으로 수정 가능.
```

wonchulee commented on internal/service/objectstorage/objectstorage_object_copy.go:
```diff
@@ -0,0 +1,296 @@
+}
+
+func (o *objectCopyResource) Update(context.Context, resource.UpdateRequest, *resource.UpdateResponse) {
+	panic("unimplemented")
```
object copy 리소스 update는 지원되지 않는 부분일까요?

Geun-Oh replied:
해당 부분 누락된 것 같습니다. 짚어주셔서 감사합니다.

Geun-Oh replied:
```
해당 서버리스 리소스들은 별도의 업데이트 로직이 발견된 것은 없어서, `panic()`구문을 삭제한 상태로 비워두겠습니다.

Object, Object Copy등등의 object storage 관련 리소스들의 경우 별도의 update 를 가지고 있지 않으며, 
aws 측에서의 구현을 보면 object에 종속되어있는 acl과 같은 요소들의 변경(타 API가 존재해서 해당 API를 호출하는 형태로 부분적으로 변경 가능한 지점)이 가능한 경우에만 변경하도록 구현되어있습니다.

따라서 현재 상태를 유지하고, 내부 attribute 중 별도 변경이 가능한 리소스가 있다면 추후 추가하겠습니다.
```

wonchulee commented on internal/service/objectstorage/objectstorage_object.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+func (o *objectResourceModel) refreshFromOutput(ctx context.Context, config *conn.ProviderConfig) {
+	output, err := config.Client.ObjectStorage.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: o.Bucket.ValueStringPointer(),
+		Key:    o.Key.ValueStringPointer(),
+	})
+	if err != nil {
+		return
+	}
+	defer output.Body.Close()
+
+	bucketName, key := TrimForParsing(o.Bucket.String()), TrimForParsing(o.Key.String())
+
+	o.ID = types.StringValue(ObjectIDGenerator(bucketName, key))
+	o.ContentLength = types.Int64PointerValue(output.ContentLength)
+	o.ContentType = types.StringPointerValue(output.ContentType)
+	o.LastModified = types.StringValue(output.LastModified.String())
+
+	bodyBytes, err := io.ReadAll(output.Body)
+	if err != nil {
+		return
+	}
+
+	o.Body = types.StringValue(string(bodyBytes))
```
업로드한 오브젝트의 크기가 큰 경우, 매번 GetObject로 값을 읽어오는 것은 부담이 될 수 있을 것으로 보입니다. HeadObject() 와같은 메타 정보를 조회하는 API 등으로 대체 가능하다면 변경하는 것이 좋겠습니다. (object_copy나, wait 하는 부분도 마찬가지)

Geun-Oh replied:
해당 부분의 경우, HeadObject에서는 실제 Body를 제공해주지 않습니다(대부분 checksum, cache 와 같은 메타 정보).
이외에도 HeadBucket, HeadObject에서는 비교적 한정적인 정보를 제공해주기 때문에 위 리소스가 제공해야하는 정보의 범위가 어느 정도인지 확인 후 변경하면 좋을 것 같습니다.

aws_s3_object 리소스에서는 Body를 optional하게나마 제공하고 있는 부분이 있어 우선 최대한의 정보를 제공할 수 있는 형태를 채택하여 구현하였습니다.

Geun-Oh replied:
https://github.com/Geun-Oh/terraform-provider-ncloud/pull/1#issuecomment-2328138634

`해당 코멘트 참고하여 정리`

wonchulee commented on internal/service/objectstorage/objectstorage_object_data_source.go:
```diff
@@ -0,0 +1,132 @@
+
+	bucketName, key := ObjectIDParser(data.ObjectID.String())
+
+	output, err := o.config.Client.ObjectStorage.GetObject(ctx, &s3.GetObjectInput{
```
(리소스와 마찬가지로 GetObject 대체 가능할지)

Geun-Oh replied:
```
해당 부분 data-source에서 body를 비롯한 meta정보를 최대한 주는 것이 어떤지에 대해 고민하다가 body를 제공할 수 있는 get 을 사용해보고자 하였습니다.

Head- 로 변경하는 대신, GetObject에서 제공하는 메타 정보들을 제공하는 코드를 추가하였습니다 :)
```

wonchulee reviewed (commented):
👍  문서나 예제는 확인하지 못했지만, 리소스/데이터 소스위주로 우선 리뷰했습니다.
전체적으로 잘 작성해주셨는데, 코멘트 남긴 몇가지 의견과 질문들 확인해주시면 좋겠습니다.
그리고 저도 objectstorage와 같이 풀네임을 사용하는 것을 선호하는데, 혹시나 사용자의 사용성 측면에서 obs와 같이 리소스 이름을 약어로 줄이는게 좋을지도 함께 한 번 생각해주시면 좋을 것 같고,
internal/service/objectstorage 아래에 위치한 리소스/데이터소스들의 파일명이 objectstorage_ prefix로 시작하는데, 해당 prefix는 제거해도 좋을 것 같습니다. (이건 사소한 것)

Geun-Oh committed:
feat: Delete panic from object_copy Update

Geun-Oh commented:
@wonchulee @youngmn 리뷰 감사합니다 :)

망 구분과 관련한 부분 및 사소한 이슈들을 처리해두겠습니다.

이외에는 원철님께서 말씀주신 것처럼 Head- API로 헤더 정보를 가져오는 것과 Get 혹은 List를 이용해서 정보를 가져오는 것의 각각의 장단이 존재하는 것 같아, 추후 어떤 리소스에서 어떤 API를 통해서 정보를 가져올지에 대한 Align이 필요할 것 같습니다.

해당 부분 하단에 정리해두겠습니당.

Geun-Oh commented:
## Bucket 및 Object 정보 관련

### Waiting resource creation & deletion

#### Bucket

HeadBucket이 최근 버킷 생성 정보를 가져오지 못하는 이슈(캐싱 이슈로 추정)가 있음.
따라서 **Bucket 생성 및 삭제 정보 Waiting => ListBuckets 필수**

이외에 Read의 경우, [HeadBucket](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3#HeadBucketOutput) 과 [ListBuckets](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3#Bucket)를 통해 얻을 수 있는 정보가 다름(메타 정보는 HeadBucket이 더 많은 정보를 줌).
=> 이 경우 HeadBucket으로 교체하면 좋을 것 같습니다.

#### Object

Object의 경우 실제 Body를 가져오기 위해서는 GetObject API를 사용해야하는데, Terraform을 이용하여 선언적인 리소스 관리가 필요한 곳에서 매번 Body 정보를 가져오는 것은 불필요하다고 판단 (HeadObject가 정보를 불러오지 못하는 등의 이슈는 X).
=> 따라서 이 부분 또한 Head로 가져오도록 변경.

---

결과적으로, wait과 관련하여 필수적으로 ListBuckets를 사용해야하는 점을 제외하고는 Head- API를 사용하여 메타 정보를 제공해주는 것으로 변경을 생각하고 있습니다.

이에, 의견 있으신 경우 코멘트 주시면 감사하겠습니다 :)

Geun-Oh committed:
feat: Remove endpoint from Config struct

Geun-Oh committed:
fix: Remove Optional keyword from creation_date(computed)

Geun-Oh committed:
refactor: set bucket unique value from bucket_id to bucket_name

Geun-Oh committed:
fix: Remove unused headbucket

Geun-Oh committed:
feat: Add bucket ID (since it uses id for default)

Geun-Oh committed:
feat: Add attributes based on HeadObjectOutput

Geun-Oh committed:
feat: Add endpoint abstract logic

Geun-Oh committed:
refactor: Move post fetch logic into refreshFromOutput

Geun-Oh committed:
feat: Add genEndpointWithCode (region, site)

Geun-Oh committed:
feat: Add optional endpoint configuration

Geun-Oh committed:
feat: Add region formatting logic

Geun-Oh committed:
feat: Add Optional variables compatible with aws sdk in object-kind resource

Geun-Oh committed:
docs: Update docs with additional attributes

Geun-Oh committed:
fix: Remove Object Encryption logic entirely

Geun-Oh committed:
feat: Add site variable in example tf code

Geun-Oh committed:
chore: Upgrade Go version to v1.21

chore: Upgrade Go version to v1.21

chore: upgrade Go to v1.21

ci: Set go version 1.21 in GitHub Actions

ci: Upgrade actions version (checkout, setup-go, ci-lint)

ci: Upgrade golangci-lint v1.59

Geun-Oh committed:
chore: Add Error with Diagnostics

Geun-Oh committed:
refactor: remove endpoint from providerConfig

Geun-Oh committed:
chore: add missed packages

Geun-Oh committed:
fix: fix endpoint env to optional

Geun-Oh committed:
fix: Parse bucket info after pulling plan data

Geun-Oh committed:
refactor: Abstract BucketNameValidator

Geun-Oh committed:
Merge branch 'main' into feat/object-storage

Geun-Oh committed:
fix: Remove optional from creation_date

Geun-Oh committed:
fix: Change serverless environment => platform independent

Geun-Oh committed:
fix: change id with object_url

Geun-Oh committed:
feat: Add attribute documentation

Geun-Oh committed:
feat: Add update logic to object_acl

Geun-Oh committed:
feat: Add update logic to bucket_acl

Geun-Oh committed:
fix: Fix typo

Revert "fix: Fix typo"

This reverts commit 37c773743b0f01e70fd675a16f3b77592117dcf2.

fix: Fix typo

Geun-Oh committed:
feat: Add mapRegionWithEnv util func

Geun-Oh committed:
docs: Add env configuration guide
